File: linkedlist/maintain_tail.js
var Node = function(val) {
    this.val = val;
    this.next = null;
};

var List = function() {

    var head = null, tail = null;

    this.__defineGetter__('head', function() { return head; });
    this.__defineGetter__('tail', function() { return tail; });

    this.prepend = function(node) {
        if (!tail) {
            tail = node;
        }
        node.next = head;
        head = node;
    };

    this.append = function(node) {
        if (!head) {
            head = node;
        }
        if (tail) {
            tail.next = node;
        }
        tail = node;
    };

    this.remove = function(node) {
        if (!node) {
            return false;
        }
        if (node === head) {
            head = node.next;
            if (node === tail) {
                tail = head;
            }
            return true;
        }
        var cur = head;
        while (cur) {
            if (cur.next === node) {
                cur.next = node.next;
                if (tail === node) {
                    tail = cur;
                }
                return true;
            }
            cur = cur.next;
        }
        return false;
    };

    this.insertAfter = function(node, val) {
        var toInsert = new Node(val);
        if (node === null) {
            this.prepend(toInsert);
            return true;
        }
        var cur = head;
        while (cur) {
            if (cur === node) {
                toInsert.next = node.next;
                node.next = toInsert;
                if (tail === node) {
                    tail = toInsert;
                }
                return true;
            }
            cur = cur.next;
        }
        return false;
    };
};

assert = require('assert');

var list, n1, n2, n3;

list = new List;
list.prepend(new Node(1));
assert.equal(list.head.val, 1);
assert.equal(list.tail.val, 1);
list.prepend(new Node(2));
assert.equal(list.head.val, 2);
assert.equal(list.tail.val, 1);
list.prepend(new Node(3));
assert.equal(list.head.val, 3);
assert.equal(list.tail.val, 1);

list = new List;
list.append(new Node(1));
assert.equal(list.head.val, 1);
assert.equal(list.tail.val, 1);
list.append(new Node(2));
assert.equal(list.head.val, 1);
assert.equal(list.tail.val, 2);
list.append(new Node(3));
assert.equal(list.head.val, 1);
assert.equal(list.tail.val, 3);

/*
 * remove tests
 */

list = new List;
assert.equal(list.remove(), false);
assert.equal(list.head, null);
assert.equal(list.tail, null);

list = new List;
assert.equal(list.remove(null), false);
assert.equal(list.head, null);
assert.equal(list.tail, null);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.remove(new Node), false);
assert.equal(list.head, n1);
assert.equal(list.tail, n3);

list = new List;
assert.equal(list.head, null);
assert.equal(list.tail, null);
n1 = new Node(1);
list.append(n1);
assert.equal(list.remove(n1), true);
assert.equal(list.head, null);
assert.equal(list.tail, null);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
list.append(n1);
list.append(n2);
assert.equal(list.remove(n2), true);
assert.equal(list.head, n1);
assert.equal(list.tail, n1);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.remove(n3), true);
assert.equal(list.head, n1);
assert.equal(list.tail, n2);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
list.append(n1);
list.append(n2);
assert.equal(list.remove(n1), true);
assert.equal(list.head, n2);
assert.equal(list.tail, n2);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.remove(n1), true);
assert.equal(list.head, n2);
assert.equal(list.tail, n3);

list = new List;
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.remove(n2), true);
assert.equal(list.head, n1);
assert.equal(list.tail, n3);

/*
 * insertAfter tests
 */

list = new List
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.insertAfter(null, 4), true);
assert.equal(list.head.val, 4);
assert.equal(list.tail, n3);

list = new List
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.insertAfter(n2, 4), true);
assert.equal(list.head, n1);
assert.equal(list.tail, n3);

list = new List
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.insertAfter(n1, 4), true);
assert.equal(list.head, n1);
assert.equal(list.tail, n3);

list = new List
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.insertAfter(n3, 4), true);
assert.equal(list.head, n1);
assert.equal(list.tail.val, 4);

list = new List
n1 = new Node(1);
n2 = new Node(2);
n3 = new Node(3);
list.append(n1);
list.append(n2);
list.append(n3);
assert.equal(list.insertAfter(new Node, 4), false);
assert.equal(list.head, n1);
assert.equal(list.tail, n3);

File: linkedlist/maintain_tail.py
class Node(object):
    def __init__(self, val):
        self.val = val
        self.next = None

class List(object):
    def __init__(self):
        self._head, self._tail = None, None

    @property
    def head(self):
        return self._head

    @property
    def tail(self):
        return self._tail

    def prepend(self, node):
        if (not self._tail):
            self._tail = node
        node.next = self._head
        self._head = node

    def append(self, node):
        if (not self._head):
            self._head = node
        if (self._tail):
            self._tail.next = node
        self._tail = node

    def remove(self, node=None):
        if not node:
            return False
        if node is self._head:
            self._head = self._head.next
            if node is self._tail:
                self._tail = self._head
            return True
        cur = self._head
        while (cur):
            if cur.next is node:
                cur.next = node.next
                if self._tail is node:
                    self._tail = cur
                return True
            cur = cur.next
        return False

    def insert_after(self, node, val):
        to_insert = Node(val)
        if not node:
            self.prepend(to_insert)
            return True
        cur = self._head
        while (cur):
            if cur is node:
                to_insert.next = cur.next
                cur.next = to_insert
                if self._tail is node:
                    self._tail = to_insert
                return True
            cur = cur.next
        return False


if __name__ == '__main__':

    list = List()
    list.prepend(Node(1))
    assert list.head.val == 1
    assert list.tail.val == 1
    list.prepend(Node(2))
    assert list.head.val == 2
    assert list.tail.val == 1
    list.prepend(Node(3))
    assert list.head.val == 3
    assert list.tail.val == 1

    list = List()
    list.append(Node(1))
    assert list.head.val == 1
    assert list.tail.val == 1
    list.append(Node(2))
    assert list.head.val == 1
    assert list.tail.val == 2
    list.append(Node(3))
    assert list.head.val == 1
    assert list.tail.val == 3

    #
    # remove tests
    #

    list = List();
    assert list.remove() is False
    assert list.head is None
    assert list.tail is None

    list = List();
    assert list.remove(None) is False
    assert list.head is None
    assert list.tail is None

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1)
    list.append(n2)
    list.append(n3)
    assert list.remove(Node) is False
    assert list.head is n1
    assert list.tail is n3

    list = List();
    assert list.head is None
    assert list.tail is None
    n1 = Node(1);
    list.append(n1)
    assert list.remove(n1) is True
    assert list.head is None
    assert list.tail is None

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    list.append(n1)
    list.append(n2)
    assert list.remove(n2) is True
    assert list.head is n1
    assert list.tail is n1

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1)
    list.append(n2)
    list.append(n3)
    assert list.remove(n3) is True
    assert list.head is n1
    assert list.tail is n2

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    list.append(n1)
    list.append(n2)
    assert list.remove(n1) is True
    assert list.head is n2
    assert list.tail is n2

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1)
    list.append(n2)
    list.append(n3)
    assert list.remove(n1) is True
    assert list.head is n2
    assert list.tail is n3

    list = List();
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1)
    list.append(n2)
    list.append(n3)
    assert list.remove(n2) is True
    assert list.head is n1
    assert list.tail is n3

    # 
    # insert_after tests
    #

    list = List()
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1);
    list.append(n2);
    list.append(n3);
    assert list.insert_after(None, 4) is True
    assert list.head.val is 4
    assert list.tail is n3

    list = List()
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1);
    list.append(n2);
    list.append(n3);
    assert list.insert_after(n2, 4) is True
    assert list.head is n1
    assert list.tail is n3

    list = List()
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1);
    list.append(n2);
    list.append(n3);
    assert list.insert_after(n1, 4) is True
    assert list.head is n1
    assert list.tail is n3

    list = List()
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1);
    list.append(n2);
    list.append(n3);
    assert list.insert_after(n3, 4) is True
    assert list.head is n1
    assert list.tail.val is 4

    list = List()
    n1 = Node(1);
    n2 = Node(2);
    n3 = Node(3);
    list.append(n1);
    list.append(n2);
    list.append(n3);
    assert list.insert_after(Node, 4) is False
    assert list.head is n1
    assert list.tail is n3
